src/sqlitestorage.cpp

```diff
@@ -1067,14 +1068,44 @@ int SqliteStorage::loadContactIncidences(const Person &person, int limit, QDateT
     }
     SL3_bind_int64(stmt1, index, secsStart);
 
-    count = d->loadIncidences(stmt1, true, limit, last, false);
+    count = d->loadIncidences(stmt1, true, nullptr, limit, last, false);
 
 error:
     d->mIsLoading = false;
 
     return count;
 }
 
+bool SqliteStorage::search(const QString &key, QStringList *identifiers)
+{
+    if (!d->mDatabase || key.isEmpty())
+        return false;
+
+    d->mIsLoading = true;
+    const char *query1 = SEARCH_COMPONENTS;
+    int qsize1 = sizeof(SEARCH_COMPONENTS);
+    const QByteArray s('%' + key.toUtf8().replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_") + '%');
+    int rv = 0;
+    sqlite3_stmt *stmt1 = NULL;
+    int index = 1;
+    Incidence::Ptr incidence;
+    QString nbook;
+    int count = -1;
+
+    qCDebug(lcMkcal) << "Searching DB for" << s;
+    SL3_prepare_v2(d->mDatabase, query1, qsize1, &stmt1, nullptr);
+    SL3_bind_text(stmt1, index, s.constData(), s.length(), SQLITE_STATIC);
+    SL3_bind_text(stmt1, index, s.constData(), s.length(), SQLITE_STATIC);
+    SL3_bind_text(stmt1, index, s.constData(), s.length(), SQLITE_STATIC);
+
+    count = d->loadIncidences(stmt1, true, identifiers);
```

Comment:
Actually is it maybe a bit strange when search() method which is defined to look up from the storage also loads some data, which it doesn't promise to do? Should this have more of selectIncidences() type of handling? Compare to deleted/modified/insertedIncidences() methods.

A bit also thinking alternatives of nemo-calendar just checking the loaded incidences, or if here there should be limits/control on how distant events should be returned. 

The ordering is now an implementation detail, but pondering should there be some control over that too.

Reply:
> Actually is it maybe a bit strange when search() method which is defined to look up from the storage also loads some data, which it doesn't promise to do?

Well, it can be discussed. I agree that it's loading more than necessary for the search, but the true results of the search that is `identifiers` are only containing matching incidences. The fact that it is loading more (it can be mentined in the documentation) is a shortcut for the main (only?) usecase:
- you want to search for incidences with `foo`, so you call `search()` which loads nothing and get the `identifiers` of incidences matching `foo`,
- the identifiers give you no information, so you need to load the incidences anyway. For each identifier, you call `loadIncidenceInstance()` which will actually load more than the instance you are interested in, by design to ensure a consistant view of the calendar.

I don't see a usecase where the search() call is not followed by a load of the instances. You may have already some of the matching identifiers in memory, so the load of these is not necessary. I agree that could be a minor optimisation.

Having the `search()` API following the current `selectIncidences()` API is not an option I think for two reasons:
- this API is broken, because it is trashing the notebook belonging. Indeed, it is returning a list of incidences, but without notebook information. The norebook information is stored (for good or bad) by a calendar holding these incidences. But since they are presented only as a list and not inside a MemoryCalendar for instance, the notebook information is lost.
- if we adjust a bit the current API to keep the notebook association (like returning a QHash<QString, Incidence::Ptr> instead of a list or returning a MemoryCalendar), the fact that we're going to use these incidences, means that we need to add them to the calendar associated to the storage. So it means to add the returned incidences to the mCalendar. So why not directly do it in the search() call ?

The problem of limit is valid. It could be added to the `search()` API, saying that we load only the first N matching incidences. Currently, I've played with it searching for stupid strings like `a`, returning 1500 matches from my database. It's not slow : around 2 seconds to get the list displayed. But I agree that it's useless to get a list of more than 20-50 matches when you look for something… So let's add a `int limit` argument that UI will hardcode to 50 or something like that.

The same for ordering, we can still add it later as an optional argument if necessary. But I don't plan to add search options to the API. Calendars are simpler than emails in my opinion and less populated. So searching with the most recent on top should cover most of the cases (like when is this doctor appointment ?).

Reply:
Yea, guess since the events would need to get loaded to show anything useful, and the loading already gets done here, better then just to add to the calendar. In that case could just consider to note on the method definition that it will load what it finds.

For the ordering, if we allow to adjust that, I guess apps would like to then also define a starting point. Otherwise if the order is reversed to show the oldest ones first, there might be quite pointless ones in the beginning.  Something like "start, end, order" might fit needs. But maybe good enough for starters with just what there is currently.